network/src/commonMain/kotlin/dev/icerock/moko/network/features/HttpRetrySendFeature.kt

```diff
@@ -0,0 +1,76 @@
+package dev.icerock.moko.network.features
+
+import io.ktor.client.HttpClient
+import io.ktor.client.features.HttpClientFeature
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.client.request.HttpSendPipeline
+import io.ktor.client.statement.HttpStatement
+import io.ktor.util.AttributeKey
+import io.ktor.utils.io.errors.IOException
+import kotlinx.coroutines.delay
+
+/**
+@param onGetDelay the callback for calculation a delay between failed request and next request
+@param maxAmountRetrying the max amounts retrying requests
+@param onConditionRetrying the callback for condition a retrying request. By default it check `e is IOException `
+ */
+class HttpRetrySendFeature(
+    private val onGetDelay: (lastDelayInMillisecond: Long, timeRetrying: Int) -> Long,
+    private val maxAmountsRetrying: Int,
+    private val onConditionRetrying: (e: Throwable) -> Boolean
+) {
+
+    class Config {
+        var onGetDelay: (lastDelayInMillisecond: Long, timeRetrying: Int) -> Long =
+            { _, _ -> 2_000L }
+        var maxAmountRetrying: Int = 3 // first request + three retrying requests,
+        var onConditionRetrying: (e: Throwable) -> Boolean = { e -> e is IOException }
+        fun build() = HttpRetrySendFeature(onGetDelay, maxAmountRetrying, onConditionRetrying)
+    }
+
+    companion object Feature : HttpClientFeature<Config, HttpRetrySendFeature> {
+        private const val LAST_DELAY_HEADER = "HttpRetrySendFeature-Last-Delay"
+        private const val RETRY_COUNTER_HEADER = "HttpRetrySendFeature-Retry-Counter"
+
+        override val key: AttributeKey<HttpRetrySendFeature> = AttributeKey("HttpRetrySendFeature")
+
+        override fun install(feature: HttpRetrySendFeature, scope: HttpClient) {
+            scope.sendPipeline.intercept(HttpSendPipeline.Before) {
+                val counter = context.headers[RETRY_COUNTER_HEADER]?.toInt() ?: 0
+                val lastDelay = context.headers[LAST_DELAY_HEADER]?.toLong()
+                    ?: feature.onGetDelay(0L, counter)
+                if (counter < feature.maxAmountsRetrying) {
+                    try {
+                        context.headers.remove(LAST_DELAY_HEADER)
+                        context.headers.remove(RETRY_COUNTER_HEADER)
+                        proceed()
```

Comment:
strange logic. we send request without headers, but later we send request with headers one time and if it failed - we not catch again...do you debug this logic?

Reply:
In final this headers not send in a futher phase after `HttpSendPipeline.Before`. Logic:
if request failed then I'm fininishing  current request(all pipeline) and create new request (new pipeline(requestPipeline->sendPipeline)) adding info in headers when new request(retryed) come in  the `HttpSendPipeline.Before` phase then I get info from headers and I delete these headers then send to next phase.
If counter more or equel maxAmountsRetrying, I send to next phase like common request, that could throw a not handled exception(in this feature).

Reply:
confused logic in code...support of this will be pain - we should rework it to more clear version. @Tetraquark will help you